pype_schema/parse_json.py

```diff
@@ -57,15 +57,102 @@ def initialize_network(self):
 
+    def collect_virtual_tags(self, config, obj_id=None, virtual_tags=None):
+        """Recursively collects all virtual tags in a network's dictionary
+        representation (i.e. config)
+
+        Parameters
+        ----------
+        config: dict
+            Dictionary representation of the network
+
+        obj_id : str
+            Optional "id" of config whose virtual tags are being collected
+            (if `None` will use "ParentNetwork")
+
+        virtual_tags : dict
+            dictionary to store virtual tags in
+
+        """
+        if virtual_tags is None:
+            virtual_tags = {}
+        obj_id = "ParentNetwork" if obj_id is None else obj_id
+        v_tags = config.get("virtual_tags")
+        if v_tags:
+            # Make sure obj_id is specified for all virtual tags as `parent_id`
+            for v_tag in v_tags.values():
+                v_tag["parent_id"] = obj_id
+            virtual_tags.update(v_tags)
+        for key, value in config.items():
+            if key not in ["tags", "virtual_tags"] and isinstance(value, dict):
+                self.collect_virtual_tags(value, obj_id=key, virtual_tags=virtual_tags)
+        return virtual_tags
+
+    def add_virtual_tags(self):
+        """Adds all virtual tags in an object
+        NOTE: assumes the objects tags have already been added
+
+        Parameters
+        ----------
+        obj : Node or Connection
+            object to add virtual tags to
+
+        config : dict
+            dictionary to search for virtual tags
+
+        node_id : str
+            optional node_id to look for virtual tags in
+
+        """
+        config_v_tags = self.collect_virtual_tags(self.config)
+        network_v_tags = {}
+        if config_v_tags:
+            # Create a queue of virtual tags to add
+            v_tag_queue = [
+                (v_tag_id, v_tag_info) for v_tag_id, v_tag_info in config_v_tags.items()
+            ]
+            counter = 0
+            while v_tag_queue:
+                v_tag_id, v_tag_info = v_tag_queue.pop(0)
+                # Try parsing the virtual tag
+                try:
+                    obj_id = v_tag_info.get("parent_id")
+                    obj = (
+                        self.network_obj
+                        if obj_id == "ParentNetwork"
+                        else self.network_obj.get_node_or_connection(
+                            obj_id, recurse=True
+                        )
+                    )
+                    v_tag = self.parse_virtual_tag(
+                        v_tag_id, v_tag_info, obj, parent_network=self.network_obj
+                    )
+                    obj.add_tag(v_tag)
+                # If there is an error skip and add other tags
+                except ValueError:
+                    v_tag_queue.append((v_tag_id, v_tag_info))
+                counter += 1
+                # Note that 2 * N is theoretical limit on number of tries to add virtual tags
```

Comment:
On second thought, it actually isn't true. I could have 1 real tag and then three virtual tags pointing in sequence that just do unit conversions. It is a good way to limit recursion depth though. I added a `max_recursion_limit=5` kwarg to the function

Reply:
Modified by removing the counter variable and doing better exception handling. If a VirtualTag initialization fails due to a KeyError, it sees whether any of the tags in its `.tags` attribute are virtual and whether they are in `config_v_tags`, if so, it just adds back to the queue, otherwise it returns a KeyError (it is referencing a nonexistent tag)